[WIP] NF: Adding pytorch versions

pjsjongsung opened the pull request:
This PR is to create pytorch versions of our tensorflow model here, and possibly in the future delete tensorflow from requirements. The PR creates a new directory dipy.nn.torch, which has the torch versions of evac and histo_resdnn. Necessary changes were made in the tensorflow model code as well, to avoid confusion.

pjsjongsung commented:
Strange. The tests pass on my system with or without pytorch. Will do some more digging

skoudoro commented on dipy/nn/histo_resdnn.py:
```diff
@@ -131,7 +131,7 @@ def load_model_weights(self, weights_path):
 
         The weights for a sh_order of 8 can be obtained via the function:
-            get_fnames('histo_resdnn_weights').
+            get_fnames('histo_resdnn_tf_weights').
```
can you add `get_fnames(name='histo_resdnn_tf_weights')`

skoudoro commented on dipy/nn/torch/evac.py:
```diff
@@ -0,0 +1,551 @@
+    def forward(self, inputs):
+        return torch.sum(
+            inputs, dim=1, keepdim=True
+        )  # Summing across channel dimension
```
why not function documentation instead comment ?

pjsjongsung replied:
It's from using llms to transfer model architectures. Unnecessary so removed.

skoudoro commented on dipy/nn/torch/evac.py:
```diff
@@ -0,0 +1,551 @@
+        self.channel_sum = ChannelSum()
+        self.add = Add()
+
+    def forward(self, x, passed):
```
is it possible to remove 1 letter parameter ? can you rename `x`?

pjsjongsung replied:
Same comment as below. I just changed it so that the input parameter wasn't x.

skoudoro commented on dipy/nn/torch/evac.py:
```diff
@@ -0,0 +1,551 @@
+        self.softmax = Softmax(dim=1)
+
+    def forward(self, inputs, raw_input_2, raw_input_3, raw_input_4, raw_input_5):
+        fwd1, x = self.block1(inputs, inputs)
```
same comment as above concerning `x`

pjsjongsung replied:
I believe in that case it will be holding in memory unnecessary intermediate results. I could change it if needed though.

skoudoro commented on dipy/nn/torch/evac.py:
```diff
@@ -0,0 +1,551 @@
+    """
+
+    @doctest_skip_parser
+    @warning_for_keywords()
```
you do not need this `@warning_for_keywords()`.  it is only for old function

skoudoro commented on dipy/nn/torch/evac.py:
```diff
@@ -0,0 +1,551 @@
+
+        # EVAC+ network load
+
+        self.model = init_model()
```
strange, I suppose it is missing a `self` here with the `init_model`

skoudoro commented on dipy/nn/torch/evac.py:
```diff
@@ -0,0 +1,551 @@
+
+    def fetch_default_weights(self):
+        """
+        Load the model pre-training weights to use for the fitting.
+        While the user can load different weights, the function
+        is mainly intended for the class function 'predict'.
```
please separate by an empty line the short and the long description

skoudoro commented on dipy/nn/torch/evac.py:
```diff
@@ -0,0 +1,551 @@
+        is mainly intended for the class function 'predict'.
+        """
+        fetch_model_weights_path = get_fnames(name="evac_default_torch_weights")
+        print(f"fetched {fetch_model_weights_path}")
```
print to remove

skoudoro commented on dipy/nn/torch/histo_resdnn.py:
```diff
@@ -0,0 +1,296 @@
+    """
+
+    @deprecated_params("sh_order", new_name="sh_order_max", since="1.9", until="2.0")
+    @warning_for_keywords()
```
you do not need this `@warning_for_keywords()`.  it is only for old function

skoudoro commented on dipy/nn/torch/histo_resdnn.py:
```diff
@@ -0,0 +1,296 @@
+    .. footbibliography::
+    """
+
+    @deprecated_params("sh_order", new_name="sh_order_max", since="1.9", until="2.0")
```
I suppose to remove since it is new

skoudoro commented on dipy/nn/torch/histo_resdnn.py:
```diff
@@ -0,0 +1,296 @@
+        self.model = DenseModel(self.sh_size, num_hidden).type(torch.float64)
+
+    def fetch_default_weights(self):
+        r"""
```
`r` not needed

skoudoro commented on dipy/nn/torch/histo_resdnn.py:
```diff
@@ -0,0 +1,296 @@
+        self.load_model_weights(fetch_model_weights_path)
+
+    def load_model_weights(self, weights_path):
+        r"""
```
same as above

skoudoro commented on dipy/nn/torch/histo_resdnn.py:
```diff
@@ -0,0 +1,296 @@
+
+        return self.model(torch.from_numpy(x_test)).detach().numpy()
+
+    @warning_for_keywords()
```
same as above

skoudoro commented on doc/doc-requirements.txt:
```diff
@@ -20,4 +20,5 @@ Jinja2
 sphinx-design>=0.5.0
 boto3
 tensorflow
+pytorch
```
can you add this in the `pyproject.toml` also. Thanks !

skoudoro reviewed (commented):
Hi @pjsjongsung,

Overall, it looks great! Below, you will see some minor comments. 
I did not see one test comparing pytorch and tensorflow version. Can you add this just in case. Thanks !

pjsjongsung commented:
Addressed most of the comments. I didn't add a torch tf comparison because the torch tests are using the same input and output as the tensorflow tests.

pjsjongsung commented:
> > Addressed most of the comments. I didn't add a torch tf comparison because the torch tests are using the same input and output as the tensorflow tests.
> 
> I see. Can you add a note somewhere to be aware of that. Thank you

Added it in the commit message. Do you prefer somewhere else, like inside a file?

skoudoro reviewed (approved):
Ok, Looks good, I am going ahead and merge this PR. Thank you for this work @pjsjongsung
